src/abstractdataframe/reshape.jl

```diff
@@ -121,137 +198,240 @@ function _stackview(df::AbstractDataFrame, measure_vars::AbstractVector{Int},
-            if !hadmissing # if it is the first time we have to add a new row
-                hadmissing = true
-                # we use the fact that missing is greater than anything
-                for i in eachindex(unstacked_val)
-                    push!(unstacked_val[i], missing)
-                end
-            end
-            i = length(unstacked_val[1])
-        else
-            i = refkref
-        end
-        if !warned_dup && mask_filled[i, kref]
-            @warn("Duplicate entries in unstack at row $k for key "*
-                  "$(refkeycol[k]) and variable $(keycol[k]).")
-            warned_dup = true
+        kref = keycol_map[keycol[k]]
+        refkref = refkeycol_map[refkeycol[k]]
+        if !allowduplicates && mask_filled[refkref, kref]
+            throw(ArgumentError("Duplicate entries in unstack at row $k for key "*
+                                "$(refkeycol[k]) and variable $(keycol[k]). " *
+                                "Pass allowduplicates=true to allow them."))
         end
-        unstacked_val[kref][i] = valuecol[k]
-        mask_filled[i, kref] = true
+        unstacked_val[kref][refkref] = valuecol[k]
+        mask_filled[refkref, kref] = true
     end
-    levs = levels(refkeycol)
-    # we have to handle a case with missings in refkeycol as levs will skip missing
-    col = similar(df[!, rowkey], length(levs) + hadmissing)
-    copyto!(col, levs)
-    hadmissing && (col[end] = missing)
-    df2 = DataFrame(unstacked_val, Symbol.(renamecols.(levels(keycol))), copycols=false)
-    return insertcols!(df2, 1, _names(df)[rowkey] => col)
+    # note that Symbol.(renamecols.(colnames)) must produce unique column names
+    # and _names(df)[rowkey] must also produce a unique name
+    df2 = DataFrame(unstacked_val, Symbol.(renamecols.(colnames)), copycols=false)
+    return insertcols!(df2, 1, _names(df)[rowkey] => col, copycols=false)
 end
 
 function unstack(df::AbstractDataFrame, rowkeys, colkey::ColumnIndex,
-                 value::ColumnIndex; renamecols::Function=identity)
+                 value::ColumnIndex; renamecols::Function=identity,
+                 allowmissing::Bool=false, allowduplicates::Bool=false)
     rowkey_ints = index(df)[rowkeys]
     @assert rowkey_ints isa AbstractVector{Int}
     length(rowkey_ints) == 0 && throw(ArgumentError("No key column found"))
     length(rowkey_ints) == 1 && return unstack(df, rowkey_ints[1], colkey, value,
-                                               renamecols=renamecols)
-    g = groupby(df, rowkey_ints, sort=true)
-    keycol = categorical(df[!, colkey])
-    droplevels!(keycol)
+                                               renamecols=renamecols,
+                                               allowmissing=allowmissing,
+                                               allowduplicates=allowduplicates)
+    local g
+    try
+        g = groupby(df, rowkey_ints, sort=true)
+    catch
```

Comment:
For unordered types, this is going to do most of the work and fail only at the end, right? So that's going to be twice slower as it could. Maybe you could pass `sort=true` only when we know it's going to be more efficient than sorting manually at the end, i.e. when the grouping column implements `refpool` (see exact conditions in `row_group_slots`; would be worth putting in a common function to keep them in sync).

Reply:
I have checked it, but I am not sure how to extract it out to make everything efficient. I proposed an alternative solution assuming https://github.com/JuliaData/DataAPI.jl/pull/26 is merged, but I am open for suggestions.

Reply:
I don't think `isordered` is relevant here: `isordered(categorical(["a", "b", "c"]))` is false yet it's the most efficient case.

Maybe the cleanest implementation is to extract these lines to two helper functions so that you can call the first one to group, and then if the second one throws you still have the result.

Or maybe we should reconsider the sorting behavior? I've just checked, and `tidyr::pivot_wider` doesn't sort: it keeps the order of appearance, contrary to its predecessor `reshape2::spread`. That kind of makes sense since if you cared about order you would have sorted the input too. So maybe we should do the same and have `sort=false` by default, so that we can simply throw an error if sorting fails when `sort=true`?

Reply:
as discussed we will stop sorting